### ISSUE TITLE
Software caused connection abort

### ISSUE THREAD

**Issue by labrusca** [MM2-0/Kvaesitso#58]
app version 1.8.1
timezone:Asia/Shanghai

Because of the GFW(Great Firewall), Some weather apis and websites are blocked in China
Is it useful infomation for you?

```
java.net.SocketException: Software caused connection abort
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:119)
	at java.net.SocketInputStream.read(SocketInputStream.java:176)
	at java.net.SocketInputStream.read(SocketInputStream.java:144)
	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.readFromSocket(ConscryptEngineSocket.java:945)
	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.processDataFromSocket(ConscryptEngineSocket.java:909)
	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.access$100(ConscryptEngineSocket.java:731)
	at com.android.org.conscrypt.ConscryptEngineSocket.doHandshake(ConscryptEngineSocket.java:241)
	at com.android.org.conscrypt.ConscryptEngineSocket.startHandshake(ConscryptEngineSocket.java:220)
	at okhttp3.internal.connection.RealConnection.connectTls(RealConnection.kt:379)
	at okhttp3.internal.connection.RealConnection.establishProtocol(RealConnection.kt:337)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.kt:209)
	at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:226)
	at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:106)
	at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:74)
	at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:255)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
	at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
	at de.mm20.launcher2.currencies.ExchangeRateWorker.doWork(ExchangeRateWorker.kt:24)
	at androidx.work.Worker$1.run(Worker.java:86)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:920)
```

**Comment by MM2-0** [comment]
Looks like the currency rates that are used for converting currencies in the unit converter cannot be updated (they are loaded from an [API provided by European Central Bank](https://www.ecb.europa.eu/stats/eurofxref/eurofxref-daily.xml)). I guess the currency converter will not work for you (type "1 USD" into the search bar to test it)

> Is it useful infomation for you?

I don't know. What am I supposed to do about this? Is it an issue if it tries to repeat this request every few hours only to fail? Or that the currency converter feature is apparently unavailable in China? I could look for another API I haven't found any alternative yet that's free to use.

**Comment by labrusca** [comment]
[API provided by European Central Bank](https://www.ecb.europa.eu/stats/eurofxref/eurofxref-daily.xml) is working well in most of time.

About the weather APIs:
- **Deutscher Wetterdienst**
Work well
- **OpenWeatherMap**
Work well but something happened
- **Net Norway**
Not work
- **HERE**
Work in sometimes


Net Norway API:
```
org.json.JSONException: Value <!DOCTYPE of type java.lang.String cannot be converted to JSONObject
	at org.json.JSON.typeMismatch(JSON.java:112)
	at org.json.JSONObject.<init>(JSONObject.java:172)
	at org.json.JSONObject.<init>(JSONObject.java:185)
	at de.mm20.launcher2.weather.metno.MetNoProvider.loadWeatherData$weather_release(MetNoProvider.kt:197)
	at de.mm20.launcher2.weather.metno.MetNoProvider.loadWeatherData$weather_release(MetNoProvider.kt:21)
	at de.mm20.launcher2.weather.LatLonWeatherProvider.loadWeatherData$suspendImpl(LatLonWeatherProvider.kt:43)
	at de.mm20.launcher2.weather.LatLonWeatherProvider.loadWeatherData$weather_release(Unknown Source:0)
	at de.mm20.launcher2.weather.WeatherProvider.fetchNewWeatherData(WeatherProvider.kt:35)
	at de.mm20.launcher2.weather.WeatherUpdateWorker.doWork(WeatherRepository.kt:237)
	at androidx.work.CoroutineWorker$startWork$1.invokeSuspend(CoroutineWorker.kt:68)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

HERE API:
```
java.net.UnknownHostException: Unable to resolve host "weather.ls.hereapi.com": No address associated with hostname
	at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:156)
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:103)
	at java.net.InetAddress.getAllByName(InetAddress.java:1152)
	at okhttp3.Dns$Companion$DnsSystem.lookup(Dns.kt:49)
	at okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress(RouteSelector.kt:164)
	at okhttp3.internal.connection.RouteSelector.nextProxy(RouteSelector.kt:129)
	at okhttp3.internal.connection.RouteSelector.next(RouteSelector.kt:71)
	at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:205)
	at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:106)
	at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:74)
	at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:255)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:920)
Caused by: android.system.GaiException: android_getaddrinfo failed: EAI_NODATA (No address associated with hostname)
	at libcore.io.Linux.android_getaddrinfo(Native Method)
	at libcore.io.ForwardingOs.android_getaddrinfo(ForwardingOs.java:136)
	at libcore.io.BlockGuardOs.android_getaddrinfo(BlockGuardOs.java:221)
	at libcore.io.ForwardingOs.android_getaddrinfo(ForwardingOs.java:136)
	at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:135)
	... 23 more
```

But it's ok, I will try to find some API available in China.

**Comment by MM2-0** [comment]
If they do work sometimes, it can't be the Great Firewall, can it? It is also possible that the request just fails because the device is not connected to the internet. As long as your weather data are up-to-date, you can simply ignore these exceptions. Maybe I should add an explanation for that. Exceptions can be helpful to debug if something isn't working as expected (e.g. if the weather widget would stop to update itself) but as long as everything is working correctly, you can ignore them.

The MET norway looks suspicious though, I might investigate that. How exactly does the GFW work? Do you simply get no response when you try to open a blocked website or do you get redirected to some kind of official error screen?

**Comment by labrusca** [comment]
>It is also possible that the request just fails because the device is not connected to the internet.

I don't think so, my phone is work well in network.

> How exactly does the GFW work? Do you simply get no response when you try to open a blocked website or do you get redirected to some kind of official error screen?

It's complicated, here is some information about GFW: [Great Firewall - Wikipedia](https://en.wikipedia.org/wiki/Great_Firewall) 
 
Sometimes, I don't even have access to github.com in my province. so I'm sure all the problems should be caused by GFW

**Comment by labrusca** [comment]
seem fixed in version 1.9.0
